File: scripts/python/test_uptime.py
"""
Integration tests for the configurable timezone introduced for `check_uptime`
(issues #365, #452, #590).

The timezone is owned per-module: each plugin caches the value of
`/settings/default/timezone` in its `loadModuleEx` (no settings-core
persistence). To switch timezone at runtime in a test we therefore have to
reload the affected module so its `loadModuleEx` runs again and re-reads the
setting.

For both `local` and `utc` we exercise `check_uptime` (CheckSystem on
Windows / CheckSystemUnix on Linux):

* the default detail-syntax surfaces the configured zone via `${tz}`;
* the `--max-unit` argument controls the granularity of `${uptime}`
  (e.g. `max-unit=d` caps the result at days);
* duration thresholds with unit suffixes (`30m`, `2h`, `1d`, `2w`)
  parse the same on both platforms (regression for #452).
"""

from NSCP import Settings, Registry, Core, log, status, log_error, sleep
from test_helper import (
    BasicTest,
    TestResult,
    install_testcases,
    init_testcases,
    shutdown_testcases,
)
import os
from time import time


# Alias used to register CheckSystem so we can reload it without affecting
# any other CheckSystem instance loaded by other tests.
SYSTEM_ALIAS = "test_uptime_system"

# CheckSystem maps to a different DLL/SO name on Windows vs. Linux.
SYSTEM_MODULE = "CheckSystem" if os.name == "nt" else "CheckSystemUnix"


class UptimeTimezoneTest(BasicTest):
    """Validate `check_uptime` under both `local` and `utc`."""

    reg = None
    conf = None
    core = None

    def desc(self):
        return "Testcase for check_uptime configurable timezone and max-unit (#365/#452/#590)"

    def title(self):
        return "check_uptime timezone tests"

    def setup(self, plugin_id, prefix):
        pass

    def teardown(self):
        pass

    # ------------------------------------------------------------------ helpers

    def _set_timezone(self, tz):
        """Write the configured timezone and reload CheckSystem so the
        per-module cache is refreshed."""
        self.conf.set_string("/settings/default", "timezone", tz)
        # The module reads the timezone in loadModuleEx; reload picks it up.
        self.core.reload(SYSTEM_ALIAS)
        # Give the module a moment to settle.
        sleep(500)

    def _query(self, args):
        return self.core.simple_query("check_uptime", args)

    # ------------------------------------------------------------ uptime tests

    def _run_uptime_for(self, tz, expected_label):
        result = TestResult("check_uptime with timezone=%s" % tz)
        self._set_timezone(tz)

        # 1. Default detail-syntax surfaces the configured zone via `(${tz})`.
        #    Override warn/crit with non-triggering thresholds so the test does
        #    not depend on how long the host has been up (the built-in defaults
        #    `warn=uptime<2d` / `crit=uptime<1d` would alert on freshly booted
        #    test machines).
        (rc, msg, _perf) = self._query(
            ["show-all", "warn=uptime < 0", "crit=uptime < 0"]
        )
        result.add_message(
            rc == status.OK,
            "default check_uptime should be OK with timezone=%s: rc=%s msg=%s"
            % (tz, rc, msg),
            "Expected OK",
        )
        result.assert_contains(
            expected_label,
            msg,
            "default detail-syntax should contain the configured tz label '%s' (got %r)"
            % (expected_label, msg),
        )
        result.assert_contains("boot:", msg, "default output should mention boot:")
        result.assert_contains("uptime:", msg, "default output should mention uptime:")

        # The opposite label must not leak through under the configured zone.
        # (Skip this assertion for the value where local==UTC could be true,
        # but treating the labels as opaque strings is fine because the
        # implementation always writes "local" or "UTC", never both.)
        opposite = "UTC" if expected_label == "local" else "local"
        result.assert_not_contains(
            opposite,
            msg,
            "configured zone is %s; opposite label '%s' should not appear (got %r)"
            % (tz, opposite, msg),
        )

        # 2. The --max-unit argument controls granularity of ${uptime} (#590).
        #    With max-unit=d, the rendered uptime must not contain a week
        #    suffix even on hosts that have been up longer than a week.
        #    Override warn/crit again to keep the result OK regardless of
        #    actual host uptime.
        (rc, msg, _perf) = self._query(
            [
                "show-all",
                "max-unit=d",
                "warn=uptime < 0",
                "crit=uptime < 0",
                "detail-syntax=uptime=${uptime} tz=${tz}",
            ]
        )
        result.add_message(
            rc == status.OK,
            "max-unit=d should resolve under timezone=%s: rc=%s msg=%s"
            % (tz, rc, msg),
            "Expected OK",
        )
        result.assert_contains(
            "uptime=",
            msg,
            "${uptime} placeholder should be substituted (got %r)" % msg,
        )
        # Reject the week suffix: max-unit=d must roll weeks into days.
        result.assert_not_contains(
            "w ",
            msg,
            "max-unit=d should not produce a week suffix (got %r)" % msg,
        )
        result.assert_contains(
            "tz=%s" % expected_label,
            msg,
            "${tz} placeholder should resolve to %r (got %r)" % (expected_label, msg),
        )

        # 2b. An invalid max-unit value should be rejected with UNKNOWN, not
        #     silently accepted.
        (rc, msg, _perf) = self._query(["show-all", "max-unit=foo"])
        result.assert_equals(
            rc,
            status.UNKNOWN,
            "max-unit=foo should fail with UNKNOWN (got rc=%s, msg=%s)" % (rc, msg),
        )

        # 3. Duration thresholds with unit suffixes must parse on both
        #    platforms (#452). The expression `uptime < 30m` previously
        #    misbehaved on Windows because the original suffix was lost.
        for spec in ["30m", "2h", "1d", "2w"]:
            (rc, msg, _perf) = self._query(
                ["show-all", "warn=uptime < %s" % spec, "detail-syntax=ok"]
            )
            result.add_message(
                rc != status.UNKNOWN,
                "uptime < %s should parse cleanly under timezone=%s (rc=%s, msg=%s)"
                % (spec, tz, rc, msg),
                "Got UNKNOWN",
            )

        return result

    # ----------------------------------------------------------------- runner

    def run_test(self, cases=None):
        master = TestResult("Testing check_uptime under local and utc")

        # Make sure the module under test is loaded before we start toggling
        # settings and reloading it.
        self.core.load_module(SYSTEM_MODULE, SYSTEM_ALIAS)

        for tz, label in [("local", "local"), ("utc", "UTC")]:
            master.add(self._run_uptime_for(tz, label))

        # Restore the default so we don't poison any subsequent test in the
        # same run.
        self._set_timezone("local")
        return master

    # ----------------------------------------------------------------- harness

    def install(self):
        self.conf.set_string("/modules", SYSTEM_ALIAS, SYSTEM_MODULE)
        self.conf.set_string("/modules", "pytest", "PythonScript")
        self.conf.set_string(
            "/settings/pytest/scripts", "test_uptime", "test_uptime.py"
        )

        # Default timezone for the run; individual test cases override.
        self.conf.set_string("/settings/default", "timezone", "local")

        self.conf.save()

    def uninstall(self):
        pass

    def help(self):
        pass

    def init(self, plugin_id):
        self.reg = Registry.get(plugin_id)
        self.conf = Settings.get(plugin_id)
        self.core = Core.get(plugin_id)

    def shutdown(self):
        pass

    def require_boot(self):
        return True


instance = UptimeTimezoneTest()
all_tests = [instance]


def __main__(args):
    install_testcases(all_tests)


def init(plugin_id, plugin_alias, script_alias):
    init_testcases(plugin_id, plugin_alias, script_alias, all_tests)


def shutdown():
    shutdown_testcases()

File: scripts/lua/test_uptime.lua
-- Protobuf-free Lua port of scripts/python/test_uptime.py.
--
-- Validates check_uptime under both `local` and `utc` timezones (issues
-- #365/#452/#590). The timezone is cached per-module in loadModuleEx, so we
-- switch it by writing /settings/default/timezone and reloading CheckSystem.
-- For each zone we assert, via core:simple_query:
--   * the default detail-syntax surfaces the configured zone label (${tz}) and
--     mentions boot:/uptime:, and the opposite label never leaks;
--   * --max-unit=d caps ${uptime} granularity (no week suffix) and resolves ${tz};
--   * an invalid --max-unit is rejected with UNKNOWN;
--   * duration thresholds with unit suffixes (30m/2h/1d/2w) parse cleanly.
-- No protobuf, no network - a CheckSystem(Unix) target under the sanitizer build.
--
-- On Linux check_uptime lives in CheckSystemUnix, whose module name is
-- "CheckSystem" (it builds as libCheckSystem.so); we load it under an alias so
-- reloads don't disturb any other CheckSystem instance.
local test = require("test_helper")

local SYSTEM_ALIAS = 'test_uptime_system'
local SYSTEM_MODULE = 'CheckSystem'

local UptimeTest = {}

function UptimeTest:install(arguments)
	local conf = Settings()
	conf:set_string('/modules', SYSTEM_ALIAS, SYSTEM_MODULE)
	conf:set_string('/modules', 'luatest', 'LUAScript')
	conf:set_string('/settings/luatest/scripts', 'test_uptime', 'test_uptime.lua')
	-- Default timezone for the run; individual cases override.
	conf:set_string('/settings/default', 'timezone', 'local')
end

function UptimeTest:setup() end

function UptimeTest:teardown()
	-- Restore the default so we don't poison subsequent tests in the same run.
	self:set_timezone('local')
end

-- Write the configured timezone and reload CheckSystem so its per-module cache
-- (populated in loadModuleEx) is refreshed.
function UptimeTest:set_timezone(tz)
	local conf = Settings()
	conf:set_string('/settings/default', 'timezone', tz)
	Core():reload(SYSTEM_ALIAS)
	nscp.sleep(500)
end

function UptimeTest:query(args)
	return Core():simple_query('check_uptime', args)
end

function UptimeTest:run_uptime_for(tz, label)
	local result = test.TestResult:new{message = 'check_uptime with timezone=' .. tz}
	self:set_timezone(tz)
	local opposite = (label == 'local') and 'UTC' or 'local'

	-- 1. Default detail-syntax surfaces the configured zone label. Override
	--    warn/crit so the result is OK regardless of actual host uptime.
	local rc, msg = self:query({'show-all', 'warn=uptime < 0', 'crit=uptime < 0'})
	result:add_message(rc == 'ok',
		string.format('default check_uptime OK with timezone=%s (rc=%s)', tz, tostring(rc)))
	result:assert_contains(msg, label, 'default detail-syntax contains tz label ' .. label)
	result:assert_contains(msg, 'boot:', 'default output mentions boot:')
	result:assert_contains(msg, 'uptime:', 'default output mentions uptime:')
	result:assert_not_contains(msg, opposite, 'opposite tz label ' .. opposite .. ' does not appear')

	-- 2. --max-unit=d caps ${uptime} granularity (no week suffix) and resolves ${tz}.
	rc, msg = self:query({'show-all', 'max-unit=d', 'warn=uptime < 0', 'crit=uptime < 0',
		'detail-syntax=uptime=${uptime} tz=${tz}'})
	result:add_message(rc == 'ok',
		string.format('max-unit=d resolves under timezone=%s (rc=%s)', tz, tostring(rc)))
	result:assert_contains(msg, 'uptime=', '${uptime} placeholder substituted')
	result:assert_not_contains(msg, 'w ', 'max-unit=d produces no week suffix')
	result:assert_contains(msg, 'tz=' .. label, '${tz} placeholder resolves to ' .. label)

	-- 2b. An invalid max-unit must be rejected with UNKNOWN.
	rc, msg = self:query({'show-all', 'max-unit=foo'})
	result:assert_equals(rc, 'unknown', 'max-unit=foo fails with UNKNOWN')

	-- 3. Duration thresholds with unit suffixes must parse cleanly (#452).
	for _, spec in ipairs({'30m', '2h', '1d', '2w'}) do
		rc, msg = self:query({'show-all', 'warn=uptime < ' .. spec, 'detail-syntax=ok'})
		result:add_message(rc ~= 'unknown',
			string.format('uptime < %s parses cleanly under timezone=%s (rc=%s)', spec, tz, tostring(rc)))
	end

	return result
end

function UptimeTest:run()
	local master = test.TestResult:new{message = 'Testing check_uptime under local and utc'}
	master:add(self:run_uptime_for('local', 'local'))
	master:add(self:run_uptime_for('utc', 'UTC'))
	return master
end

local instances = { UptimeTest }
test.init_test_manager(instances)

function main(args)
	return test.install_test_manager(instances)
end
